File: Tests/RedlineCoreTests/FindingsDismissalTests.swift
// A dismissal has to hide a finding everywhere at once, and has to expire. Both are the point.
import XCTest
@testable import RedlineCore

final class FindingsDismissalTests: XCTestCase {
    private let now = Date(timeIntervalSince1970: 1_787_000_000)

    private func finding(_ id: String, kind: Finding.Kind = .fixNow) -> Finding {
        Finding(id: id, kind: kind, basis: .measured, title: id, detail: "d", evidence: [])
    }

    private func report(_ ids: [String]) -> FindingsReport {
        FindingsReport(generatedAt: now, windowDays: 14, sessionsScanned: 3,
                       findings: ids.map { finding($0) })
    }

    func testDismissedFindingIsHiddenUntilTheSnoozeExpires() {
        var d = FindingsDismissals()
        d.dismiss("mcp-unused", at: now)
        XCTAssertTrue(d.isHidden("mcp-unused", snoozeDays: 14, now: now))
        // A day short of the window is still hidden; a day past it is not
        XCTAssertTrue(d.isHidden("mcp-unused", snoozeDays: 14,
                                 now: now.addingTimeInterval(13 * 86400)))
        XCTAssertFalse(d.isHidden("mcp-unused", snoozeDays: 14,
                                  now: now.addingTimeInterval(15 * 86400)))
        XCTAssertFalse(d.isHidden("reread-files", snoozeDays: 14, now: now))
    }

    /// The menu line and the panel both read `visible`, so this is what keeps them agreeing
    func testVisibleFiltersAndCountsWhatItHid() {
        var d = FindingsDismissals()
        d.dismiss("a", at: now)
        let v = report(["a", "b", "c"]).visible(d, snoozeDays: 14, now: now)
        XCTAssertEqual(v.findings.map(\.id), ["b", "c"])
        XCTAssertEqual(v.hidden, 1)
        XCTAssertEqual(v.summary, "2 findings · 2 to fix")
        // Everything else about the report survives the filter
        XCTAssertEqual(v.sessionsScanned, 3)
        XCTAssertEqual(v.windowDays, 14)
    }

    /// A finding that is still true after the snooze must come back, not vanish quietly
    func testAStillTrueFindingReturnsAfterTheSnooze() {
        var d = FindingsDismissals()
        d.dismiss("a", at: now)
        let later = now.addingTimeInterval(20 * 86400)
        let v = report(["a", "b"]).visible(d, snoozeDays: 14, now: later)
        XCTAssertEqual(v.findings.map(\.id), ["a", "b"])
        XCTAssertEqual(v.hidden, 0)
    }

    func testEmptyAfterDismissingEverythingStillReportsWhatItHid() {
        var d = FindingsDismissals()
        d.dismiss("a", at: now)
        d.dismiss("b", at: now)
        let v = report(["a", "b"]).visible(d, snoozeDays: 14, now: now)
        XCTAssertTrue(v.isEmpty)
        XCTAssertEqual(v.hidden, 2)
        XCTAssertEqual(v.summary, "no findings")
    }

    func testRestoreAllBringsThemBackNow() {
        var d = FindingsDismissals()
        d.dismiss("a", at: now)
        d.restoreAll()
        XCTAssertTrue(d.isEmpty)
        XCTAssertEqual(report(["a"]).visible(d, snoozeDays: 14, now: now).findings.count, 1)
    }

    /// The file must not accumulate a row for every finding the checks ever produced
    func testPruneDropsExpiredSnoozesOnly() {
        var d = FindingsDismissals()
        d.dismiss("old", at: now.addingTimeInterval(-30 * 86400))
        d.dismiss("fresh", at: now)
        d.prune(snoozeDays: 14, now: now)
        XCTAssertFalse(d.isHidden("old", snoozeDays: 14, now: now))
        XCTAssertTrue(d.isHidden("fresh", snoozeDays: 14, now: now))
        XCTAssertEqual(d.dismissed.count, 1)
    }

    /// A clock that moved backwards must not read as a snooze that has already run out
    func testAFutureDismissalIsTreatedAsHidden() {
        var d = FindingsDismissals()
        d.dismiss("a", at: now.addingTimeInterval(86400))
        XCTAssertTrue(d.isHidden("a", snoozeDays: 14, now: now))
    }

    func testRoundTripsThroughDiskAndSurvivesAMissingFile() throws {
        let dir = URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent("redline-dismiss-" + UUID().uuidString)
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        defer { try? FileManager.default.removeItem(at: dir) }
        let url = dir.appendingPathComponent("findings-dismissed.json")

        // Absent is empty, not an error
        XCTAssertTrue(FindingsDismissalStore.load(from: url).isEmpty)

        var d = FindingsDismissals()
        d.dismiss("mcp-unused", at: now)
        XCTAssertTrue(FindingsDismissalStore.save(d, to: url))
        XCTAssertEqual(FindingsDismissalStore.load(from: url), d)

        // Garbage on disk degrades to empty rather than taking the panel out
        try "not json".write(to: url, atomically: true, encoding: .utf8)
        XCTAssertTrue(FindingsDismissalStore.load(from: url).isEmpty)
    }

    func testSnoozeDaysIsValidatedAndOutOfRangeFallsBack() {
        XCTAssertEqual(Config().findingsSnoozeDays, 14)
        XCTAssertEqual(Config.apply(["findingsSnoozeDays": 30], to: Config())
                        .findingsSnoozeDays, 30)
        for bad: Any in [0, -1, 400, "sometimes"] {
            XCTAssertEqual(Config.apply(["findingsSnoozeDays": bad], to: Config())
                            .findingsSnoozeDays, 14, "\(bad) should have been rejected")
        }
    }
}

File: windows/tests/Redline.Core.Tests/FindingsDismissalTests.cs
// A dismissal has to hide a finding everywhere at once, and has to expire. Both are the point.
using System.Text.Json.Nodes;
using Redline.Core;

namespace Redline.Core.Tests;

public sealed class FindingsDismissalTests
{
    private static readonly DateTimeOffset Now = DateTimeOffset.FromUnixTimeSeconds(1_787_000_000);

    private static Finding Finding(string id, FindingKind kind = FindingKind.FixNow) =>
        new(id, kind, FindingBasis.Measured, id, "d", Array.Empty<FindingEvidence>());

    private static FindingsReport Report(params string[] ids) =>
        new(Now, 14, 3, ids.Select(id => Finding(id)).ToList());

    [Fact]
    public void DismissedFindingIsHiddenUntilTheSnoozeExpires()
    {
        var d = new FindingsDismissals();
        d.Dismiss("mcp-unused", Now);
        Assert.True(d.IsHidden("mcp-unused", 14, Now));
        // A day short of the window is still hidden; a day past it is not
        Assert.True(d.IsHidden("mcp-unused", 14, Now.AddSeconds(13 * 86400)));
        Assert.False(d.IsHidden("mcp-unused", 14, Now.AddSeconds(15 * 86400)));
        Assert.False(d.IsHidden("reread-files", 14, Now));
    }

    /// The menu line and the panel both read `Visible`, so this is what keeps them agreeing
    [Fact]
    public void VisibleFiltersAndCountsWhatItHid()
    {
        var d = new FindingsDismissals();
        d.Dismiss("a", Now);
        var v = Report("a", "b", "c").Visible(d, 14, Now);
        Assert.Equal(new[] { "b", "c" }, v.Findings.Select(f => f.Id));
        Assert.Equal(1, v.Hidden);
        Assert.Equal("2 findings · 2 to fix", v.Summary);
        // Everything else about the report survives the filter
        Assert.Equal(3, v.SessionsScanned);
        Assert.Equal(14, v.WindowDays);
    }

    /// A finding that is still true after the snooze must come back, not vanish quietly
    [Fact]
    public void AStillTrueFindingReturnsAfterTheSnooze()
    {
        var d = new FindingsDismissals();
        d.Dismiss("a", Now);
        var v = Report("a", "b").Visible(d, 14, Now.AddSeconds(20 * 86400));
        Assert.Equal(new[] { "a", "b" }, v.Findings.Select(f => f.Id));
        Assert.Equal(0, v.Hidden);
    }

    [Fact]
    public void EmptyAfterDismissingEverythingStillReportsWhatItHid()
    {
        var d = new FindingsDismissals();
        d.Dismiss("a", Now);
        d.Dismiss("b", Now);
        var v = Report("a", "b").Visible(d, 14, Now);
        Assert.True(v.IsEmpty);
        Assert.Equal(2, v.Hidden);
        Assert.Equal("no findings", v.Summary);
    }

    [Fact]
    public void RestoreAllBringsThemBackNow()
    {
        var d = new FindingsDismissals();
        d.Dismiss("a", Now);
        d.RestoreAll();
        Assert.True(d.IsEmpty);
        Assert.Single(Report("a").Visible(d, 14, Now).Findings);
    }

    /// The file must not accumulate a row for every finding the checks ever produced
    [Fact]
    public void PruneDropsExpiredSnoozesOnly()
    {
        var d = new FindingsDismissals();
        d.Dismiss("old", Now.AddSeconds(-30 * 86400));
        d.Dismiss("fresh", Now);
        d.Prune(14, Now);
        Assert.False(d.IsHidden("old", 14, Now));
        Assert.True(d.IsHidden("fresh", 14, Now));
        Assert.Single(d.Dismissed);
    }

    /// A clock that moved backwards must not read as a snooze that has already run out
    [Fact]
    public void AFutureDismissalIsTreatedAsHidden()
    {
        var d = new FindingsDismissals();
        d.Dismiss("a", Now.AddSeconds(86400));
        Assert.True(d.IsHidden("a", 14, Now));
    }

    [Fact]
    public void RoundTripsThroughDiskAndSurvivesAMissingFile()
    {
        var dir = Path.Combine(Path.GetTempPath(), "redline-dismiss-" + Guid.NewGuid());
        Directory.CreateDirectory(dir);
        try
        {
            var path = Path.Combine(dir, "findings-dismissed.json");
            // Absent is empty, not an error
            Assert.True(FindingsDismissalStore.Load(path).IsEmpty);

            var d = new FindingsDismissals();
            d.Dismiss("mcp-unused", Now);
            Assert.True(FindingsDismissalStore.Save(d, path));
            Assert.Equal(d, FindingsDismissalStore.Load(path));

            // Garbage on disk degrades to empty rather than taking the panel out
            File.WriteAllText(path, "not json");
            Assert.True(FindingsDismissalStore.Load(path).IsEmpty);
        }
        finally { try { Directory.Delete(dir, true); } catch { } }
    }

    [Fact]
    public void SnoozeDaysIsValidatedAndOutOfRangeFallsBack()
    {
        Assert.Equal(14, new Config().FindingsSnoozeDays);
        var ok = JsonNode.Parse("{\"findingsSnoozeDays\": 30}")!.AsObject();
        Assert.Equal(30, Config.Apply(ok, new Config()).FindingsSnoozeDays);
        foreach (var bad in new[] { "0", "-1", "400", "\"sometimes\"" })
        {
            var json = JsonNode.Parse("{\"findingsSnoozeDays\": " + bad + "}")!.AsObject();
            Assert.True(Config.Apply(json, new Config()).FindingsSnoozeDays == 14,
                $"{bad} should have been rejected");
        }
    }
}
